### ISSUE TITLE
Bug: Input Secrets must be created before Azure Resources

### ISSUE THREAD

**Issue by Samze** [Azure/azure-service-operator#2385]
**Version of Azure Service Operator**: mcr.microsoft.com/k8s/azureserviceoperator:v2.0.0-beta.1

**Describe the bug**
Azure Operator does not re-reconcile on missing Secret. This means the Secret *Must* exist before the Resource, this makes doing things like gitops hard where ordering is not guaranteed.

**To Reproduce**
1. Create a flexible server
```yaml
cat <<EOF | kubectl apply -f -
apiVersion: dbforpostgresql.azure.com/v1alpha1api20210601
kind: FlexibleServer
metadata:
  name: aso-psql-bug
spec:
  location: centralus
  owner:
    name: aso-psql-rg
  version: "13"
  sku:
    name: Standard_D4s_v3
    tier: GeneralPurpose
  administratorLogin: myAdmin
  administratorLoginPassword:
    name: aso-psql-bug
    key: password
  storage:
    storageSizeGB: 128
EOF
```

2. Observe the correct error in status
```yaml
status:
  conditions:
  - lastTransitionTime: "2022-07-14T21:35:07Z"
    message: 'failed resolving secret references: default/aso-psql-bug does not exist
      (Secret "aso-psql-bug" not found)'
    observedGeneration: 1
    reason: SecretNotFound
    severity: Warning
    status: "False"
    type: Ready
```

3. Now create the secret
```yaml
cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: Secret
metadata:
  annotations:
  name: aso-psql-bug
  namespace: default
data:
  password: Zm9vCg==
type: Opaque
EOF
---
```

4. Watch the FlexibleServer and observe its status never changes from the error.

**Expected behavior**
A missing Secret should be treated as an error and the reconciler should retry with some backoff and it eventual succeed.

**Comment by matthchr** [comment]
Thanks for this report. We totally agree. Will fix this for the next release.

**Comment by matthchr** [comment]
I looked into this and it turns out that we are already doing the right thing here:

1. Your resource had condition `Ready = False` with `severity = Warning`. A severity of `Warning` means that the operator will continue to attempt to reconcile that resource over and over. So even if we didn't have an event triggered when the secret was created, eventually a retry on reconciliation would figure it out and proceed.
2. We _do_ correctly have an event that triggers when a secret is created that immediately triggers reconciliation for the resource that secret belongs to. I confirmed this both in our test environment and a "real" ASO deployment

That doesn't mean you're not seeing a problem here though, you are. I believe this is probably what is happening:
1. Initial creation of the DB fails because the secret isn't there.
2. You add the secret. This allows the resource creation to proceed, but importantly the `Condition` saying there was a problem with the secret is not yet overwritten (more on this later).
3. Resource creation runs into bug #2372, where it fails but thinks it passes, which leaves the resource in a bad state. This also isn't setting the `Condition`.

`Warning` conditions are not overwritten by `Info` conditions with the same `ObservedGeneration`. This _normally_ is the correct behavior but in the case of a secrets update it's not. It means that you can end up in a situation where you're still seeing a warning that you've fixed. You'll continue to see that warning until something updates the `Generation` of the resource (such as modifying the `Spec`) or until the resource hits a different Warning, Error, or succeeds provisioning.

So the fix here is twofold. First, we need to fix #2372, which is already fixed and will be part of v2-beta.2. Second, we need to make sure that warnings related to secrets _are_ overwritten by subsequent `Info` conditions.
